File: src/main/java/com/dingdangmaoup/resin/pura/ui/DeploymentSettingsEditor.java
package com.dingdangmaoup.resin.pura.ui;

import com.dingdangmaoup.resin.pura.ResinModuleDeploymentModel;
import com.intellij.javaee.appServers.deployment.DeploymentModel;
import com.intellij.javaee.appServers.deployment.DeploymentSource;
import com.intellij.javaee.appServers.run.configuration.CommonModel;
import com.intellij.openapi.options.ConfigurationException;
import com.intellij.openapi.options.SettingsEditor;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import javax.swing.JCheckBox;
import javax.swing.JComponent;
import javax.swing.JPanel;
import javax.swing.JTextField;
import org.jetbrains.annotations.NotNull;

public class DeploymentSettingsEditor extends SettingsEditor<DeploymentModel> {

  private JPanel myMainPanel;

  private JTextField myHostField;
  private JTextField myApplicationContextField;
  private JCheckBox myDefaultContextCheckBox;

  public DeploymentSettingsEditor(final CommonModel commonModel, final DeploymentSource deploymentSource) {
    super(() -> new ResinModuleDeploymentModel(commonModel, deploymentSource));
    myDefaultContextCheckBox.addActionListener(new ActionListener() {
      @Override
      public void actionPerformed(ActionEvent e) {
        updateContextEnabled();
      }
    });
    myDefaultContextCheckBox.setSelected(true);
    updateContextEnabled();
  }

  private void updateContextEnabled() {
    myApplicationContextField.setEnabled(!myDefaultContextCheckBox.isSelected());
  }

  @Override
  public void resetEditorFrom(@NotNull DeploymentModel settings) {
    ResinModuleDeploymentModel resinDeploymentModel = (ResinModuleDeploymentModel)settings;
    myDefaultContextCheckBox.setSelected(resinDeploymentModel.isDefaultContextPath());
    updateContextEnabled();
    myApplicationContextField.setText(resinDeploymentModel.getContextPath());
    myHostField.setText(resinDeploymentModel.getHost());
  }

  @Override
  public void applyEditorTo(@NotNull DeploymentModel settings) throws ConfigurationException {
    ResinModuleDeploymentModel resinDeploymentModel = (ResinModuleDeploymentModel)settings;
    resinDeploymentModel.setDefaultContextPath(myDefaultContextCheckBox.isSelected());
    resinDeploymentModel.setContextPath(myApplicationContextField.getText());
    resinDeploymentModel.setHost(myHostField.getText());
  }

  @Override
  @NotNull
  protected JComponent createEditor() {
    return myMainPanel;
  }
}

File: src/main/kotlin/com/dingdangmaoup/resin/pura/ui/DeploymentSettingsEditor.kt
package com.dingdangmaoup.resin.pura.ui

import com.dingdangmaoup.resin.pura.ResinBundle
import com.dingdangmaoup.resin.pura.ResinModuleDeploymentModel
import com.intellij.javaee.appServers.deployment.DeploymentModel
import com.intellij.javaee.appServers.deployment.DeploymentSource
import com.intellij.javaee.appServers.run.configuration.CommonModel
import com.intellij.openapi.options.ConfigurationException
import com.intellij.openapi.options.SettingsEditor
import com.intellij.ui.dsl.builder.AlignX
import com.intellij.ui.dsl.builder.panel
import javax.swing.JCheckBox
import javax.swing.JComponent
import javax.swing.JLabel
import javax.swing.JTextField

class DeploymentSettingsEditor(
    commonModel: CommonModel,
    deploymentSource: DeploymentSource,
) : SettingsEditor<DeploymentModel>({ ResinModuleDeploymentModel(commonModel, deploymentSource) }) {
    private val myHostField = JTextField(28)
    private val myApplicationContextField = JTextField(28)
    private val myDefaultContextCheckBox = JCheckBox(ResinBundle.message("deployment.dlg.use.default.context.path"))
    private val myContextPathLabel = JLabel(ResinBundle.message("deployment.dlg.context.path"))
    private val myHostLabel = JLabel(ResinBundle.message("deployment.dlg.host"))
    private val myMainPanel = panel {
        row {
            cell(myDefaultContextCheckBox)
        }
        row {
            cell(myContextPathLabel)
            cell(myApplicationContextField).align(AlignX.FILL)
        }
        row {
            cell(myHostLabel)
            cell(myHostField).align(AlignX.FILL)
        }
    }

    init {
        myDefaultContextCheckBox.addActionListener { updateContextEnabled() }
        myDefaultContextCheckBox.isSelected = true
        updateContextEnabled()
    }

    private fun updateContextEnabled() {
        myApplicationContextField.isEnabled = !myDefaultContextCheckBox.isSelected
    }

    override fun resetEditorFrom(settings: DeploymentModel) {
        val resinDeploymentModel = settings as ResinModuleDeploymentModel
        myDefaultContextCheckBox.isSelected = resinDeploymentModel.isDefaultContextPath
        updateContextEnabled()
        myApplicationContextField.text = resinDeploymentModel.contextPath
        myHostField.text = resinDeploymentModel.host
    }

    @Throws(ConfigurationException::class)
    override fun applyEditorTo(settings: DeploymentModel) {
        val resinDeploymentModel = settings as ResinModuleDeploymentModel
        resinDeploymentModel.isDefaultContextPath = myDefaultContextCheckBox.isSelected
        resinDeploymentModel.contextPath = myApplicationContextField.text
        resinDeploymentModel.host = myHostField.text
    }

    override fun createEditor(): JComponent = myMainPanel
}
